2D AFPM example

ngadiyar93 opened the pull request:
This PR implements a 2D AFPM Example which:
1) Draws the AFPM equivalent 2D Model
2) Assigns Boundary Condition
3) Solves the model in MagNet (Transient)
4) Extracts results from MagNet into Matlab
5) Plots the results

ngadiyar93 commented:
This PR is ready for review.

elsevers commented on tools/magnet/plumbing/document/mn_d_createEvenPeriodicBoundaryCondition.m:
```diff
@@ -0,0 +1,39 @@
+function [boundary] = mn_d_createEvenPeriodicBoundaryCondition(mn, problemID, surfacesFace, Face , x, y, z, name )
```
`mn_d_createEvenPeriodicBoundaryCondition()` should be a porcelain function -- please move it over to the porcelain location.

elsevers commented on tools/magnet/plumbing/document/mn_d_createEvenPeriodicBoundaryCondition.m:
```diff
@@ -0,0 +1,39 @@
+        invoke(mn, 'processcommand', ...
+        sprintf('nmArray(0)= "%s,Face#%i"', surfacesFace, Face));    
+    end
+cmdstring = sprintf('call getDocument.createBoundaryCondition(nmArray,"%s")', name); 
```
Create a plumbing function for `createBoundaryCondition`

elsevers commented on tools/magnet/plumbing/document/mn_d_createEvenPeriodicBoundaryCondition.m:
```diff
@@ -0,0 +1,39 @@
+invoke(mn, 'processcommand', sprintf('ShiftVector(0)= %s', num2str(x)));
+invoke(mn, 'processcommand', sprintf('ShiftVector(1)= %s', num2str(y)));
+invoke(mn, 'processcommand', sprintf('ShiftVector(2)= %s', num2str(z)));
+invoke(mn, 'processcommand', sprintf('call getDocument.setEvenPeriodic("%s",Null,Null,Null,Null,ShiftVector,Null)',name));
```
Create a plumbing function for `setEvenPeriodic`

elsevers commented on tools/magnet/plumbing/document/solution/mn_getCoilFlux.m:
```diff
@@ -0,0 +1,27 @@
+function dataPoints = mn_getCoilFlux(mn, doc, compName, pid)
```
We already have the function `mn_readCoilFluxLinkage()`. Please take a look -- if the existing function is sufficient, let's not add this new function.

elsevers commented on tools/magnet/plumbing/document/solution/mn_getCoilVoltage.m:
```diff
@@ -0,0 +1,27 @@
+function dataPoints = mn_getCoilVoltage(mn, doc, compName, pid)
```
`mn_readCoilVoltage()` already exists.

elsevers commented on tools/magnet/plumbing/document/solution/mn_getComponentForce.m:
```diff
@@ -0,0 +1,28 @@
+function dataPoints = mn_getComponentForce(mn, doc, compName, pid, axis)
```
We already have `mn_readForceOnBody()`

ngadiyar93 commented:
I have made the changes suggested and the PR is ready for review.

elsevers commented on tools/magnet/plumbing/document/mn_d_setEvenPeriodic.m:
```diff
@@ -0,0 +1,13 @@
+invoke(mn, 'processcommand', sprintf('ShiftVector(0)= %s', num2str(x)));
+invoke(mn, 'processcommand', sprintf('ShiftVector(1)= %s', num2str(y)));
+invoke(mn, 'processcommand', sprintf('ShiftVector(2)= %s', num2str(z)));
+invoke(mn, 'processcommand', sprintf('call getDocument.setEvenPeriodic("%s",Null,Null,Null,Null,ShiftVector,Null)',name));
```
Since we are adding this as a plumbing function, the pattern to follow is to make your argument list as close to the actual functions argument list as possible. This means you should make the following changes:
1. You should not have `x, y, z` as separate arguments. Rather, they should be one array (`shiftVector`?), as listed in the MagNet `setEvenPeriodic` function
2. It appears that you have only implemented a couple of the arguments of `setEvenPeriodic()`. Please add support for all arguments (I am referring to all of the nulls). If this is not possible, then make your function compatible with the full argument list so that functionality can be added later.
3. Does MagNet's `setEvenPeriodic()` return a value? If so, it should be returned from this function.

ngadiyar93 replied:
I have made these changes. 
MagNet's setEvenPeriodic() doesn't return a value.

elsevers commented on tools/magnet/porcelain/mn_d_createEvenPeriodicBoundaryCondition.m:
```diff
@@ -0,0 +1,39 @@
+        invoke(mn, 'processcommand', ...
+        sprintf('nmArray(0)= "%s,Face#%i"', surfacesFace, Face));    
+    end
+cmdstring = sprintf('call getDocument.createBoundaryCondition(nmArray,"%s")', name); 
```
This should be a call to the function you have created `mn_d_createBoundaryCondition()` instead of calling it directly here

elsevers commented on tools/magnet/porcelain/mn_d_createEvenPeriodicBoundaryCondition.m:
```diff
@@ -0,0 +1,39 @@
+invoke(mn, 'processcommand', sprintf('ShiftVector(0)= %s', num2str(x)));
+invoke(mn, 'processcommand', sprintf('ShiftVector(1)= %s', num2str(y)));
+invoke(mn, 'processcommand', sprintf('ShiftVector(2)= %s', num2str(z)));
+invoke(mn, 'processcommand', sprintf('call getDocument.setEvenPeriodic("%s",Null,Null,Null,Null,ShiftVector,Null)',name));
```
This should be a call to the function you have created `mn_d_setEvenPeriodic()` instead of calling MagNet's function directly here

elsevers commented on tools/magnet/plumbing/document/mn_d_makeMotionComponent.m:
```diff
@@ -0,0 +1,28 @@
+function [motion] = mn_d_makeMotionComponent(mn, problemID, surfacesVolume)
```
Why is `problemID` an argument? It seems that this is not used in this function...

elsevers commented on tools/magnet/plumbing/document/mn_d_setEvenPeriodic.m:
```diff
@@ -0,0 +1,54 @@
+%   ShiftVector: coordinates of the shift vector in (mm)
+%   Center: reference point coordinates in (mm)
+%   This is a wrapper for Document::setEvenPeriodic
+if strcmp(ShiftVector,'Null')== 0
```
Instead of the user passing in a string of `Null`, the user should pass in an empty array `[ ]`. You can then check this by using the `isempty()` function. So line 12 would be:

`if ~isempty(shiftVector)`

Please apply this to the other vectors which can be null.

ngadiyar93 commented on tools/magnet/plumbing/document/mn_d_setEvenPeriodic.m:
```diff
@@ -0,0 +1,54 @@
+    reference = 'Null';
+end
+
+arguments = sprintf('"%s",%s,%s,%s, %s, %s, %s',...
```
I am not sure how this happened. Most likely I hadn't pushed once I made the update, as the version here has a line missing which actually makes the settings in MagNet. 
I have corrected this now and tested it works.

The test for now would be running the TransientSolveAFPM2DExample.m file located in `examples/exampleAFPM2D` which draws the machine, sets the conditions and solves it.

ngadiyar93 commented:
> I have found several problems. Please address my concerns.
> 
> Moreover, you need to take a careful look at your code. Please check for whitespace errors (unnecessary spaces, missing spaces, etc.), variable names that do not follow the eMach style guide or do not accurately reflect the variable they are describing, inaccurate or incomplete comments, and MagNet functions that do not match the MagNet documentation for the function they are wrapping.
> 
> Please test your code before requesting another review of your PR.

ngadiyar93 commented:
> I have found several problems. Please address my concerns.
> 
> Moreover, you need to take a careful look at your code. Please check for whitespace errors (unnecessary spaces, missing spaces, etc.), variable names that do not follow the eMach style guide or do not accurately reflect the variable they are describing, inaccurate or incomplete comments, and MagNet functions that do not match the MagNet documentation for the function they are wrapping.
> 
> Please test your code before requesting another review of your PR.

I have implemented the changes

elsevers commented on tools/magnet/plumbing/document/mn_d_makeMotionComponent.m:
```diff
@@ -0,0 +1,26 @@
+%       geometry to be moved. 
+%
+%   Return value: a path to the motion component.
+%   This is a wrapper for Document::makeMotionComponent
```
This function's comments are not consistent with our style.

Please identify standard function comment style. If we don't have it properly documented in the [knowledgebase](https://github.com/Severson-Group/KnowledgeBase/blob/master/code/matlab/README.md#style-guidelines) please create a PR in the knowledgebase to add the necessary documentation.

elsevers commented on tools/magnet/plumbing/document/mn_d_createBoundaryCondition.m:
```diff
@@ -0,0 +1,31 @@
+%       the format {compName,Face#i} where compName is the name of the
+%       component and Face#i is the face number to which the boundary
+%       condition is applied. i is an integer denoting face number.
+%   name: name assigned to the boundary condition
```
Why does name have a `:` after it?

elsevers commented on tools/magnet/plumbing/document/mn_d_createBoundaryCondition.m:
```diff
@@ -0,0 +1,31 @@
+%       component and Face#i is the face number to which the boundary
+%       condition is applied. i is an integer denoting face number.
+%   name: name assigned to the boundary condition
+%   This is a wrapper for Document::createBoundaryCondition
```
Where is the info on what is returned?

elsevers commented on tools/magnet/MagNet.m:
```diff
@@ -63,6 +63,13 @@ function close(obj)
+        function setCores(obj, numCores)
+            %SETCORES Sets the number of cores
+            %setCores(numCores) sets the numCores number of cores
+            cores = sprintf('Call getDocument().setNumberOfMultiCoreSolveThreads(%i)',numCores);           
```
There should be an empty line between the end of the description and the first line of code

elsevers reviewed (changes requested):
I have reviewed everything except for your changes in the `examples` folder. I

 have identified several items that need to be resolved. Please see my inline comments.

Please take a look through the changes you are proposing in the `examples` folder to make sure it will pass my review.

When you are ready for me to review your code again, please re-request my review.

ngadiyar93 commented on tools/magnet/porcelain/mn_d_createEvenPeriodicBoundaryCondition.m:
```diff
@@ -0,0 +1,19 @@
+%   This is a porcelain function that uses mn_d_createBoundaryCondition and
+%   mn_d_setEvenPeriodic to create even periodic boundary condition
+mn_d_createBoundaryCondition(mn, componentNames, face, name );
+mn_d_setEvenPeriodic(mn, name, scale_factor, rotation_angle, RotationAxis, MirrorNormal, ShiftVector, Center);
```
This was a redundant function I think I originally created. Yes, it doesn't make sense to have it anymore. Even the example doesn't use this - it uses the plumbing commands.

Deleting this..

elsevers commented:
@ngadiyar93 I went through the broader repository and I agree that our comment/documentation issue is widespread. Let's address the documentation issues in a separate PR. If you feel you have addressed my other requests and you are happy with the `examples` content please re-request my review. Thanks.

ngadiyar93 commented:
> @ngadiyar93 I went through the broader repository and I agree that our comment/documentation issue is widespread. Let's address the documentation issues in a separate PR. If you feel you have addressed my other requests and you are happy with the `examples` content please re-request my review. Thanks.

The PR can be re-reviewed. I have addressed the issues and also tweaked content in the examples folder. I have saved documentation updates for another PR .

elsevers commented on examples/exampleAFPM2D/ConstructAFPM2DExample.m:
```diff
@@ -0,0 +1,553 @@
+    compMag2e = Component( ...
+        'name', 'mag2e', ...
+        'crossSections', csmagnet2e, ...
+        'material', MaterialGeneric('name', 'Recoma 33E;Type=Uniform;Direction=[0,-1,0]'), ... % Recoma 35 E 
```
Why is there a comment here that says Recoma 35E?

ngadiyar93 replied:
@elsevers Oops, this was an error that I missed fixing. It had to be Recoma 33E. I have updated it now.

elsevers reviewed (changes requested):
@ngadiyar93 -- nice work. This is shaping up to be a terrific example! 

I have one request/question that I would like you to address (flagged in your code). Apart from this, I have made some corrections directly to the code for you to merge in.

Upon completing this, I am ready to approve your PR. Please collect one more review from the current eMach team.

ngadiyar93 commented:
> @ngadiyar93 -- nice work. This is shaping up to be a terrific example!
> 
> I have one request/question that I would like you to address (flagged in your code). Apart from this, I have made some corrections directly to the code for you to merge in.
> 
> Upon completing this, I am ready to approve your PR. Please collect one more review from the current eMach team.

@elsevers I have merged the patch, and also updated the comment to reflect the appropriate PM material.

elsevers commented on examples/exampleAFPM2D/ConstructAFPM2DExample.m:
```diff
@@ -0,0 +1,553 @@
+compMag2e = Component( ...
+        'name', 'mag2e', ...
+        'crossSections', csmagnet2e, ...
+        'material', MaterialGeneric('name', 'Recoma 33E;Type=Uniform;Direction=[0,-1,0]'), ... % Recoma 33 E 
```
@ngadiyar93 I think you should remove this comment. 

I don't understand why this line of code should have a `Recoma 33 E` comment when other lines that assign the magnet do not. I also don't see why it is necessary given the that material is current self-described (`Recoma 33E` is the name of the material)

ngadiyar93 replied:
@elsevers I have removed the `Recoma 33E` comments

elsevers reviewed (approved):
Nice work @ngadiyar93! I am ready to approve now.

Bharat-Ramadas reviewed (approved):
Works fine

ngadiyar93 commented:
@elsevers @Bharat-Ramadas thanks for your reviews. I just hit the merge button.
